File: tests/data-conversion/undefined/undefined.cpp
#include "runtime/object.h"
#include "runtime/union.h"

namespace {

class LinkNode : public compilets::Object {
 public:
  compilets::Union<std::monostate, double, bool> item;
  cppgc::Member<LinkNode> next;

  void Trace(cppgc::Visitor* visitor) const override {
    compilets::TraceMember(visitor, next);
  }

  virtual ~LinkNode() = default;
};

void TestUndefined() {
  std::nullopt_t undef = std::nullopt;
  std::optional<double> orUndefined = 123;
  orUndefined = std::nullopt;
  orUndefined = std::nullopt;
  compilets::Union<compilets::Null, double> orNull;
  orNull = compilets::Null{};
  compilets::Union<std::monostate, double, bool> optionalUnion;
  optionalUnion = std::monostate{};
  LinkNode* node = compilets::MakeObject<LinkNode>();
  node->item = true;
  node->next = compilets::MakeObject<LinkNode>();
  node->next = nullptr;
}

}  // namespace

File: tests/data-conversion/undefined/undefined.ts
class LinkNode {
  item: number | boolean | undefined;
  next: LinkNode | undefined;
}

function TestUndefined() {
  let undef = undefined;
  let orUndefined: number | undefined = 123;
  orUndefined = undefined;
  orUndefined = undef;
  let orNull: number | null;
  orNull = null;
  let optionalUnion: number | boolean | undefined;
  optionalUnion = undefined;

  let node = new LinkNode;
  node.item = true;
  node.next = new LinkNode;
  node.next = undefined;
}
